kubeadm: add docs for UpgradeAddonsBeforeControlPlane feature gate (#41606)
kubernetes/website: content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md

`PublicKeysECDSA` | `false` | 1.19 | - | -
`RootlessControlPlane` | `false` | 1.22 | - | -
`UnversionedKubeletConfigMap` | `true` | 1.22 | 1.23 | 1.25

> `UpgradeAddonsBeforeControlPlane` | `false` | - | - | -

Author: Thanks @tengqm 
Sorry for the late reply.

It isn't a new feature, it's more like a bugfix. So this feature gate has been marked deprecated since it was created.
So it's not in Alpha stage, and will never be Beta or GA.
For more context: https://github.com/kubernetes/kubernetes/pull/116570#discussion_r1179352015

> `UpgradeAddonsBeforeControlPlane` | `false` | - | - | -

Reviewer: In that case, we should document this in the second table (gates for GA'ed or deprecated feature).

> `UpgradeAddonsBeforeControlPlane` | `false` | - | - | -

Author: OK, will update soon.

(during `kubeadm join`, `kubeadm reset`, `kubeadm upgrade ...`), kubeadm attempts to use unversioned ConfigMap name first;
if that does not succeed, kubeadm falls back to using the legacy (versioned) name for that ConfigMap.

`UpgradeAddonsBeforeControlPlane`

> : This flag is introduced as a **deprecated** feature gate in v1.28. Before v1.27 (including v1.27), `kubeadm` will upgrade 

Reviewer 2: ```suggestion
: This flag is introduced as a **deprecated** feature gate in v1.28. Before v1.27 (including v1.27), kubeadm will upgrade 
```

if it's the component name, kubeadm should not be in code quotes

> : This flag is introduced as a **deprecated** feature gate in v1.28. Before v1.27 (including v1.27), `kubeadm` will upgrade 

Reviewer 2: ```suggestion
:  introduced as a **deprecated** feature gate in v1.28. Before v1.27 (including v1.27), `kubeadm` will upgrade 
```

perhaps omit "flag"

addons (including CoreDNS and kube-proxy) immediately during `kubeadm upgrade apply`, regardless of whether there are other 

> control plane instances that have not been upgraded, which may cause compatibility problems. Since v1.28, kubeadm will 

Reviewer 3: ```suggestion
control plane instances that have not been upgraded, which may cause compatibility problems. Since v1.28, kubeadm defaults to a mode that
```

always checks whether all the control plane instances have been upgraded before starting to upgrade the addons, this behavior 
is applied to both `kubeadm upgrade apply` and `kubeadm upgrade node`. kubeadm determines whether a control plane instance 

> has been upgraded by checking whether the image of kube-apiserver Pod has been upgraded. You **MUST** perform control plane 

Reviewer 2: ```suggestion
has been upgraded by checking whether the image of the kube-apiserver Pod has been upgraded. You **MUST** perform control plane 
```

> has been upgraded by checking whether the image of `kube-apiserver` Pod has been upgraded. You **MUST** perform control plane 

Reviewer 2: ```suggestion
has been upgraded by checking whether the image of kube-apiserver Pod has been upgraded. You **MUST** perform control plane 
```

nodes upgrade sequentially, and the addons will be upgraded after the last control plane has been upgraded. The deprecated 

> `UpgradeAddonsBeforeControlPlane` feature gate gives you a chance to keep the old upgrade behavior, but please do not rely 

Reviewer 3: ```suggestion
`UpgradeAddonsBeforeControlPlane` feature gate gives you a chance to keep the old upgrade behavior.
You should not need this old behavior; if you do, you should change your cluster or upgrade processes
so that you do not rely
```

> on it, and it will be removed in a future release.

Reviewer 2: ```suggestion
on it, as it will be removed in a future release.
```

if that does not succeed, kubeadm falls back to using the legacy (versioned) name for that ConfigMap.

`UpgradeAddonsBeforeControlPlane`
: This is as a **disabled** feature gate that was introduced for Kubernetes v1.28, in order to allow reactivating a legacy
and deprecated behavior during cluster upgrade. For kubeadm versions prior to v1.28, kubeadm upgrades cluster addons (including
CoreDNS and kube-proxy) immediately during `kubeadm upgrade apply`, regardless of whether there are other control plane
instances that have not been upgraded, which may cause compatibility problems. Since v1.28, kubeadm defaults to a mode that
always checks whether all the control plane instances have been upgraded before starting to upgrade the addons, this behavior
is applied to both `kubeadm upgrade apply` and `kubeadm upgrade node`. kubeadm determines whether a control plane instance
has been upgraded by checking whether the image of the kube-apiserver Pod has been upgraded. You **MUST** perform control plane
instances upgrade sequentially or at least ensure that the last control plane instance is upgraded after all the other control

> plane instances have been upgraded, and the addons upgrade will be performed after the last control plane instance is upgraded.

Reviewer: Ensure "that the last control plane instance is upgraded after all the other control
plane instances have been upgraded" ...

This is so verbose. Does this mean that we need to ensure all control plane instances are properly upgraded?

Author: This is because the control plane node upgrade needs to ensure the sequence. In a HA cluster, the upgrade of the last control plane node cannot be performed until all the other control plane nodes have been upgraded successfully.

> The deprecated `UpgradeAddonsBeforeControlPlane` feature gate gives you a chance to keep the old upgrade behavior. You should not

Reviewer: Is the gate defaulted to true or false?
It is introduced as a deprecated feature, right?

Author: Yes, the deprecated feature gate is defaulted to false.

> need this old behavior; if you do, you should change your cluster or upgrade processes but please do not rely on it, as this

Reviewer: "please do not rely on it" ...
What does this "it" refer to?

Author: "it" means the `UpgradeAddonsBeforeControlPlane` feature gate.

Author: "please do not rely on it" has been removed now
